### PR TITLE
protocols: Add Block I/O protocol

### PR DESCRIPTION
**Pull request by makubacki**

### CODE REVIEW

**Comment by makubacki** [comment]
Hello! As Bret mentioned in #20 we're upstreaming a few changes here.

I noticed you have an AUTHORS file with a COPYRIGHT section. For all of my PRs, could you please help me understand if I should add an Apache copyright header with the company copyright (as Bret did in his PR) or if the copyright would be added to that file?

File: src/protocols/block_io.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    pub media_id: u32,
+    pub removable_media: bool,
+    pub media_present: bool,
+    pub logical_partition: bool,
+    pub read_only: bool,
+    pub write_caching: bool,
+    pub block_size: u32,
+    pub io_align: u32,
+    pub last_block: u64,
+    pub lowest_aligned_lba: u64,
```
Comment (dvdhrm) [inline src/protocols/block_io.rs:32]: These could use `crate::base::Lba`.

File: src/protocols/block_io.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    pub logical_blocks_per_physical_block: u64,
```
Comment (dvdhrm) [inline src/protocols/block_io.rs:33]: This should be `u32` according to the spec.

File: src/protocols/block_io.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    &[0x0, 0xa0, 0xc9, 0x69, 0x72, 0x3b],
+);
+
+pub const REVISION: u64 = 0x0000000000010000u64;
+pub const REVISION2: u64 = 0x0000000000020001u64;
+pub const REVISION3: u64 = 0x000000000002001fu64;
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct BlockIoMedia {
```
Comment (dvdhrm) [inline src/protocols/block_io.rs:22]: Similar to the disk-io protocols, I would drop the `BlockIo` prefix here, as it is already part of the module-name.

**Review by dvdhrm — COMMENTED** [review]
I fix some copy errors from the spec. I hope I didn't screw them up. Let me know if I did.

Thanks a lot for the PR!

**Comment by dvdhrm** [comment]
> Hello! As Bret mentioned in #20 we're upstreaming a few changes here.

Thanks a lot for this! I appreciate this a lot!

Just as a note: I try to keep the `r-efi` crate "policy free". This is why I am a bit pedantic at times regarding naming and symbol placement. But I think this base crate should just be an import of the specification definitions, with as little deviations as possible. I hope all my C to Rust conversions are clear to follow. If not, let me know and I will try to formalize them better.

We at Red Hat did intend to publish more crates under the `r-efi` github organization, but so far have not settled on any particular helper crates other than the `r-efi-string` and `r-efi-alloc` crates.

> I noticed you have an AUTHORS file with a COPYRIGHT section. For all of my PRs, could you please help me understand if I should add an Apache copyright header with the company copyright (as Bret did in his PR) or if the copyright would be added to that file?

Yeah, I usually list all authors by name in `AUTHORS`. Regarding copyright, I prefer listing them there as well. Unless you object, I will gladly add `2019-2020 Microsoft Corporation` to the list.

I am generally not a fan of putting license info into each file. But if a contributor wishes, I might not object ;) However, in the case of r-efi, it is really just a matter of copying information out of the specification, so I would probably just list all the contributors and copyrights in `AUTHORS` and leave the individual code files as clean as possible.


Thanks for the PR! I picked up the patch and applied a correction on top.

**Comment by makubacki** [comment]
I agree with just putting the copyright in AUTHORS. Please do add the Microsoft copyright there.

All your edits looked good. I appreciate your prompt review and feedback. Thanks!

**Comment by dvdhrm** [comment]
> I agree with just putting the copyright in AUTHORS. Please do add the Microsoft copyright there.
> 
> All your edits looked good. I appreciate your prompt review and feedback. Thanks!

I've updated the AUTHORS file. Thanks for the patches!

**Pull request closed without merging**